File: src/control.rs
use biosphere::{MaxFeatures, RandomForestParameters};

/// Storage container for hyperparameters.
#[derive(Clone)]
pub struct Control {
    /// Segments with length smaller than `2 * n * minimal_relative_segment_length` will
    /// not be split.
    pub minimal_relative_segment_length: f64,
    /// Only keep split point if the gain exceeds `minimal_gain_to_split`. Relevant for
    /// change in mean. Note that this is relative to the number of observations.
    /// Use value motivated by BIC `minimal_gain_to_split = log(n_samples) * n_features / n_samples`
    /// if `None`.
    pub minimal_gain_to_split: Option<f64>,
    /// Type two error in model selection to be approximated. Relevant for classifier-
    /// based changepoint detection.
    pub model_selection_alpha: f64,
    /// Number of permutations for model selection in classifier-based change point
    /// detection.
    pub model_selection_n_permutations: usize,
    /// Number of randomly drawn segments. Corresponds to parameter `M` in
    /// https://arxiv.org/pdf/1411.0858.pdf.
    pub number_of_wild_segments: usize,
    /// Decay parameter in seeded binary segmentation. Should be in `[1/2, 1)`, with a
    /// value close to 1 resulting in many segments. Corresponds to `\alpha` in
    /// https://arxiv.org/pdf/2002.06633.pdf.
    pub seeded_segments_alpha: f64,
    /// Seed used for segmentation.
    pub seed: u64,
    /// Hyperparameters for random forests.
    pub random_forest_parameters: RandomForestParameters,
    /// Segments of indexes were no segmentation is allowed.
    pub forbidden_segments: Option<Vec<(usize, usize)>>,
}

impl Control {
    #[allow(clippy::should_implement_trait)]
    pub fn default() -> Control {
        Control {
            minimal_relative_segment_length: 0.01,
            minimal_gain_to_split: None,
            model_selection_alpha: 0.02,
            model_selection_n_permutations: 199,
            number_of_wild_segments: 100,
            seeded_segments_alpha: std::f64::consts::FRAC_1_SQRT_2, // 1 / sqrt(2)
            seed: 0,
            random_forest_parameters: RandomForestParameters::default()
                .with_max_depth(Some(8))
                .with_max_features(MaxFeatures::Sqrt)
                .with_n_jobs(Some(-1)),
            forbidden_segments: None,
        }
    }

    pub fn with_minimal_relative_segment_length(
        mut self,
        minimal_relative_segment_length: f64,
    ) -> Self {
        if (minimal_relative_segment_length >= 0.5) | (minimal_relative_segment_length <= 0.) {
            panic!(
                "minimal_relative_segment_length needs to be strictly between 0 and 0.5 Got {minimal_relative_segment_length}"
            );
        }
        self.minimal_relative_segment_length = minimal_relative_segment_length;
        self
    }

    pub fn with_minimal_gain_to_split(mut self, minimal_gain_to_split: Option<f64>) -> Self {
        self.minimal_gain_to_split = minimal_gain_to_split;
        self
    }

    pub fn with_model_selection_alpha(mut self, model_selection_alpha: f64) -> Self {
        if (model_selection_alpha >= 1.) | (model_selection_alpha <= 0.) {
            panic!(
                "model_selection_alpha needs to be strictly between 0 and 1. Got {model_selection_alpha}"                
            );
        }
        self.model_selection_alpha = model_selection_alpha;
        self
    }

    pub fn with_model_selection_n_permutations(
        mut self,
        model_selection_n_permutations: usize,
    ) -> Self {
        self.model_selection_n_permutations = model_selection_n_permutations;
        self
    }

    pub fn with_number_of_wild_segments(mut self, number_of_wild_segments: usize) -> Self {
        self.number_of_wild_segments = number_of_wild_segments;
        self
    }

    pub fn with_seeded_segments_alpha(mut self, seeded_segments_alpha: f64) -> Self {
        if (1. <= seeded_segments_alpha) | (seeded_segments_alpha <= 0.) {
            panic!(
                "seeded_segments_alpha needs to be strictly between 0 and 1. Got {seeded_segments_alpha}"                
            );
        }
        self.seeded_segments_alpha = seeded_segments_alpha;
        self
    }

    pub fn with_seed(mut self, seed: u64) -> Self {
        self.seed = seed;
        self
    }

    pub fn with_random_forest_parameters(
        mut self,
        random_forest_parameters: RandomForestParameters,
    ) -> Self {
        self.random_forest_parameters = random_forest_parameters;
        self
    }

    pub fn with_forbidden_segments(
        mut self,
        forbidden_segments: Option<Vec<(usize, usize)>>,
    ) -> Self {
        // check that segments are well specified
        if let Some(ref _forbidden_segments) = forbidden_segments {
            for el in _forbidden_segments.iter() {
                if el.0 > el.1 {
                    panic!("Forbidden segments must be specified as [(a,b), ...] where a <= b!");
                }
            }
        }
        self.forbidden_segments = forbidden_segments;
        self
    }
}

File: changeforest-r/R/control.R
#' Storage container for hyperparameters.
#'
#' See rust documentation of changeforest::control::Control,
#' biosphere::forest::RandomForestParameters and biosphere::tree::DecisionTreeParameters
#' for more details.
#'
#' @param minimal_relative_segment_length Segments with length smaller than
#' \code{2 * n * minimal_relative_segment_length} will not be split. Equal to 0.01 by default.
#' @param minimal_gain_to_split Only keep split point if the gain exceeds
#' \code{minimal_gain_to_split}. Relevant for change in mean. Use value motivated by BIC
#' \code{minimal_gain_to_split = log(n_samples) * n_features / n_samples} by default.
#' @param model_selection_alpha Type two error in model selection to be approximated.
#' Relevant for classifier-based changepoint detection. Equal to 0.02 by default.
#' @param model_selection_n_permutations Number of permutations for model selection in
#' classifier-based change point detection. Equal to 199 by default.
#' @param number_of_wild_segments Number of randomly drawn segments. Corresponds to
#' parameter \code{M} in https://arxiv.org/pdf/1411.0858.pdf. Only relevant if
#' \code{segmentation='wbs'}. Equal to 100 by default.
#' @param seeded_segments_alpha Decay parameter in seeded binary segmentation. Should
#' be in \code{[1/2, 1)}, with a value close to 1 resulting in many segments. Corresponds to
#' \eqn{\alpha} in https://arxiv.org/pdf/2002.06633.pdf. Only relevant if
#' \code{segmentatin='sbs'}. Equal to \eqn{1 / \sqrt{2}} by default.
#' @param seed Seed for segmentation and random forest. Only relevant for
#' \code{segmentation='wbs'} or \code{method='random_forest'}.
#' @param random_forest_n_estimators Parameter passed to random forest classifier if
#' \code{method='random_forest'}. Equal to 100 by default.
#' @param random_forest_max_features Parameter passed to random forest classifier if
#' \code{method='random_forest'}. Equal to \eqn{\sqrt{d}} by default.
#' @param random_forest_max_depth Parameter passed to random forest classifier if
#' \code{method='random_forest'}. Equal to 8 by default.
#' @param random_forest_n_jobs Parameter passed to random forest classifier if
#' \code{method='random_forest'}. Use all cores if -1. Equal to -1 by default.
#'
#' @return Object of class Control containing hyperparameters.
#' @export
Control = R6::R6Class(
    "control",
    list(
        #' @field minimal_relative_segment_length Segments with length smaller than 
        #' \code{2 * n * minimal_relative_segment_length} will not be split. Equal to
        #' 0.01 by default.
        minimal_relative_segment_length = "default",
        #' @field minimal_gain_to_split Only keep split point if the gain exceeds
        #' \code{minimal_gain_to_split}. Relevant for change in mean. Use value motivated by BIC
        #' \code{minimal_gain_to_split = log(n_samples) * n_features / n_samples} by default.
        minimal_gain_to_split = "default",
        #' @field model_selection_alpha Type two error in model selection to be approximated.
        #' Relevant for classifier-based changepoint detection. Equal to 0.02 by default.
        model_selection_alpha = "default",
        #' @field model_selection_n_permutations Number of permutations for model selection in
        #' classifier-based change point detection. Equal to 199 by default.
        model_selection_n_permutations = "default",
        #' @field number_of_wild_segments Number of randomly drawn segments. Corresponds to
        #' parameter \code{M} in https://arxiv.org/pdf/1411.0858.pdf. Only relevant if
        #' \code{segmentation='wbs'}. Equal to 100 by default.
        number_of_wild_segments = "default",
        #' @field seeded_segments_alpha Decay parameter in seeded binary segmentation. Should
        #' be in \code{[1/2, 1)}, with a value close to 1 resulting in many segments. Corresponds to
        #' \eqn{\alpha} in https://arxiv.org/pdf/2002.06633.pdf. Only relevant if
        #' \code{segmentatin='sbs'}. Equal to \eqn{1 / \sqrt{2}} by default.
        seeded_segments_alpha = "default",
        #' @field seed Seed for segmentation and random forest. Only relevant for
        #' \code{segmentation='wbs'} or \code{method='random_forest'}.
        seed = "default",
        #' @field random_forest_n_estimators Parameter passed to random forest classifier if
        #' \code{method='random_forest'}. Equal to 100 by default.
        random_forest_n_estimators = "default",
        #' @field random_forest_max_features Parameter passed to random forest classifier if
        #' \code{method='random_forest'}. Equal to \eqn{\sqrt{d}} by default.
        random_forest_max_features = "default",    
        #' @field random_forest_max_depth Parameter passed to random forest classifier if
        #' \code{method='random_forest'}. Equal to 8 by default. 
        random_forest_max_depth = "default",
        #' @field random_forest_n_jobs Parameter passed to random forest classifier if
        #' \code{method='random_forest'}. Use all cores if -1. Equal to -1 by default.
        random_forest_n_jobs = "default",

        #' @description
        #' Create a new object of class \code{binary_segmentation_resutl}.
        #' @param minimal_relative_segment_length Segments with length smaller than
        #' \code{2 * n * minimal_relative_segment_length} will not be split. Equal to 0.01 by default.
        #' @param minimal_gain_to_split Only keep split point if the gain exceeds
        #' \code{minimal_gain_to_split}. Relevant for change in mean. Use value motivated by BIC
        #' \code{minimal_gain_to_split = log(n_samples) * n_features / n_samples} by default.
        #' @param model_selection_alpha Type two error in model selection to be approximated.
        #' Relevant for classifier-based changepoint detection. Equal to 0.02 by default.
        #' @param model_selection_n_permutations Number of permutations for model selection in
        #' classifier-based change point detection. Equal to 199 by default.
        #' @param number_of_wild_segments Number of randomly drawn segments. Corresponds to
        #' parameter \code{M} in https://arxiv.org/pdf/1411.0858.pdf. Only relevant if
        #' \code{segmentation='wbs'}. Equal to 100 by default.
        #' @param seeded_segments_alpha Decay parameter in seeded binary segmentation. Should
        #' be in \code{[1/2, 1)}, with a value close to 1 resulting in many segments. Corresponds to
        #' \eqn{\alpha} in https://arxiv.org/pdf/2002.06633.pdf. Only relevant if
        #' \code{segmentatin='sbs'}. Equal to \eqn{1 / \sqrt{2}} by default.
        #' @param seed Seed for segmentation and random forest. Only relevant for
        #' \code{segmentation='wbs'} or \code{method='random_forest'}.
        #' @param random_forest_n_estimators Parameter passed to random forest classifier if
        #' \code{method='random_forest'}. Equal to 100 by default.
        #' @param random_forest_max_features Parameter passed to random forest classifier if
        #' \code{method='random_forest'}. Equal to \eqn{\sqrt{d}} by default.
        #' @param random_forest_max_depth Parameter passed to random forest classifier if
        #' \code{method='random_forest'}. Equal to 8 by default.
        #' @param random_forest_n_jobs Parameter passed to random forest classifier if
        #' \code{method='random_forest'}. Use all cores if -1. Equal to -1 by default.
        #' @return A new object of class \code{binary_segmentation_resutl}.
        initialize = function(
            minimal_relative_segment_length = "default",
            minimal_gain_to_split = "default",
            model_selection_alpha = "default",
            model_selection_n_permutations = "default",
            number_of_wild_segments = "default",
            seeded_segments_alpha = "default",
            seed = "default",
            random_forest_n_estimators = "default",
            random_forest_max_features = "default",
            random_forest_max_depth = "default",
            random_forest_n_jobs = "default"
        ) {
            self$minimal_relative_segment_length = minimal_relative_segment_length
            self$minimal_gain_to_split = minimal_gain_to_split
            self$model_selection_alpha = model_selection_alpha
            self$model_selection_n_permutations = model_selection_n_permutations
            self$number_of_wild_segments = number_of_wild_segments
            self$seeded_segments_alpha = seeded_segments_alpha
            self$seed = seed
            self$random_forest_n_estimators = random_forest_n_estimators 
            self$random_forest_max_features = random_forest_max_features
            self$random_forest_max_depth = random_forest_max_depth
            self$random_forest_n_jobs = random_forest_n_jobs
        }
    )
)
